feat: migrate from CRA4 to Vite, Storybook to v8, jest to Vitest

karrui opened the pull request:
This PR moves the frontend toolchain from CRA4 to Vite. 
Generally everything seems to be working fine, including decryption workers.

Requires a bunch of changes that has to be performed all at the same time:
- [x] Jest -> Vitest
- [x] Storybook v6 -> v8 (less painful than wrangling Storybook v6 to work with vite)


# TODO:
- [x] move back dd-chunk pipeline to vite build
- [ ] Test on GSIB

karrui committed:
feat: add initial vite config and new eslint rules

karrui committed:
feat: replace CRA variables with Vite variables

karrui committed:
feat: update ts-config

karrui committed:
feat: remove craco, fix webpack -> esbuild errors

karrui committed:
feat: update vite config for proxy and node polyfills

karrui committed:
feat: make decryption workers work on Vite

karrui committed:
fix: theme typing generation script

karrui committed:
fix: bundle worker as es

karrui commented:
* **#7308** <a href="https://app.graphite.dev/github/pr/opengovsg/FormSG/7308?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7305** <a href="https://app.graphite.dev/github/pr/opengovsg/FormSG/7305?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7292** <a href="https://app.graphite.dev/github/pr/opengovsg/FormSG/7292?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7285** <a href="https://app.graphite.dev/github/pr/opengovsg/FormSG/7285?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#7279** <a href="https://app.graphite.dev/github/pr/opengovsg/FormSG/7279?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7278** <a href="https://app.graphite.dev/github/pr/opengovsg/FormSG/7278?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7277** <a href="https://app.graphite.dev/github/pr/opengovsg/FormSG/7277?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @karrui and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

karrui committed:
fix: update storybook to work with vite

karrui committed:
feat: upgrade to storybook v8 package with vite builder

karrui committed:
fix: remove deprecated turbo-build storybook addon

karrui committed:
fix: add missing node-stdlib-browser package

karrui commented:
@KenLSM see chromatic useful. will ensure the migration has 0 frontend changes!!!

karrui committed:
feat: switch to vitest, fix storybook test invocations

karrui committed:
feat: format and lint

karrui committed:
feat: update msw package (to v1 latest, not v2 yet)

v2 requires more changes, not worth it for now

karrui committed:
fix: set svgr loader to not inject default dimensions

karrui committed:
fix: regression in storybook 7 that removes 100% height from stories

karrui committed:
fix: remove preview changes from `.storybook/main.ts`

actually good to snapshot full page instead of just fixed pages

karrui committed:
fix(test): use expect instead of just awaiting canvas.findByText

karrui committed:
fix: flakey toast test due to multiple tests running at the same time

hypothesis is that multiple toasts are rendering???

karrui committed:
fix: modal bounding box for chromatic stories

karrui commented:
@justynoh this PR officially has no regressions hehehehe

karrui committed:
feat: reuse old github secret env var instead of renaming to VITE_APP

karrui committed:
feat: remove unused .builtime-env

does not seem to be needed now that we use vite

karrui committed:
fix: misrenamed import

karrui committed:
feat: load datadog-chunk as separate script from main react app

karrui commented:
@justynoh added datadog chunking back to the pipeline. Can push to staging to try?

karrui commented on frontend/package.json:
```diff
@@ -102,20 +106,18 @@
-    "test": "cross-env SKIP_PREFLIGHT_CHECK=true craco test --silent",
+    "start": "vite",
+    "build": "vite build",
+    "serve": "vite preview",
```
can remove don't add confusion

KenLSM replied:
2856457124b1c2e7c0d27697e82661f9df658916

karrui commented on frontend/package.json:
```diff
@@ -102,20 +106,18 @@
+    "build": "vite build",
+    "serve": "vite preview",
+    "test": "cross-env SKIP_PREFLIGHT_CHECK=true vitest",
     "eject": "react-scripts eject",
```
no need this anymore

KenLSM replied:
2856457124b1c2e7c0d27697e82661f9df658916

karrui commented on frontend/package.json:
```diff
@@ -102,20 +106,18 @@
     "eject": "react-scripts eject",
-    "storybook": "NODE_OPTIONS=--openssl-legacy-provider start-storybook -p 6006",
-    "build-storybook": "NODE_OPTIONS=--openssl-legacy-provider build-storybook",
+    "storybook": "NODE_OPTIONS=--openssl-legacy-provider storybook dev -p 6006",
```
no need node options anymore

KenLSM replied:
2856457124b1c2e7c0d27697e82661f9df658916

KenLSM committed:
chore: remove unused cmds, update cmds

karrui committed:
feat: upgrade to react 18 and chakra v2 (#7292)

* feat: upgrade react and chakra-ui
react v17 -> react v18
chakra v1 -> chakra v2

* fix: breaking changes from Chakra v2

* fix: use React 18

* feat: update framer-motion, MotionBox instantiation

* feat: update changed React.FC to FCC prop type

* fix: sync all type-fest package types

* fix: remove unnecessary package patches

* feat: remove unnecessary scripts and script comments

* fix: correctly generate chakra theme types

* fix: revert motionbox back to old implementation

new implementation doesn't work lol

* fix: upgrade serverless type-fest to sync with rest of app

* fix: use updated useToken call signature

* fix: update react-joyride for React 18 compat

* fix: types for allowed MyInfo types

* fix: broken types assertions or code raised by linter

* fix: invalid React SVG prop clip-path -> clipPath

* fix: replace dnd package with React 18 compatible package

* fix: update package to remove console warning about deprecated calls

ReactMarkdown: Support for defaultProps will be removed from function components in a future major release.

* fix: temporary fix for menu focus color desync

final fix is to move to OGPDS

* feat: update `@testing-library/*` packages

remove unused testing-library/dom package

* feat: lock storybook packages

* fix: correctly upgrade virus-scanner type-fest package

* fix: use `isDisabled` prop over disabled

will not correctly set state if isDisabled is used instead

* fix: test assertions due to RTL changes

* fix: upgrade floating-ui package for compatibility

* fix: compat issues in ChakraUI v2

* feat: update input theming for ChakraUI v2

* fix: NumberInput render

* feat: update Drawer and Modal themes to have background using cssVars

* fix: use explicit style context required by Chakra V2

* fix: correctly import from component Button

* fix: update margin due to flex using `gap` now

instead of the margin-top previously, which allowed for margin collapse

* fix: update Rating component spacing due to flex gap change in chakra

* fix: update YesNo field margins due to Flex gap change

* fix: weird margins with collaborator menu, match width

* fix: use __css instead of sx for FeatureBanner

or text will be space-betweened

* fix: button spacing for CreateWorkspaceModal

* fix: button spacing for DeleteWorkspaceModal

* fix: button spacing for RenameWorkspaceModal

* fix: spacing on LandingPage sections

* feat: update ParagraphField story to catch multiline changes too

* feat: remove unnecessary focus-visible package

* feat: update `isTruncated` -> `noOfLines`

* feat: fix width of PermissionDropdown for consistent widths

* fix: align ViewOnlyPermission row margins

* fix: SingleSelect theme breaking because of Menu

really need to go to OGPDS soon lol

* fix: add padding to prefill lock icon

* fix: add correct msw handlers for workspace page

* fix: add back missing margins

* fix: make AvatarMenu have automatic height

instead of following button's height

* fix: add back missing menuitem padding for SingleSelect

* fix: update imported lottie file type

* fix: correct modal story color scheme for cancel button

* fix: set minH of TagInput to input's var

* fix: use input exported css variables for styling TagInput

* fix: move combobox input to the top when focused

so the border won't look janky

* feat: use mockdate decorator so skeleton width fixed in snapshot

* fix: spacing of collaborator list view again

* fix: landing page margins

* fix: avatarmenudivider margins

* fix: remove removed patches copy in dockerfile

* remove ineffective white bg on style header

---------

Co-authored-by: Ken <ken@open.gov.sg>

karrui committed:
feat: upgrade typescript to 5.4.5 across app (#7305)

* feat: upgrade react and chakra-ui
react v17 -> react v18
chakra v1 -> chakra v2

* fix: breaking changes from Chakra v2

* fix: use React 18

* feat: update framer-motion, MotionBox instantiation

* feat: update changed React.FC to FCC prop type

* fix: sync all type-fest package types

* fix: remove unnecessary package patches

* feat: remove unnecessary scripts and script comments

* fix: correctly generate chakra theme types

* fix: revert motionbox back to old implementation

new implementation doesn't work lol

* fix: upgrade serverless type-fest to sync with rest of app

* fix: use updated useToken call signature

* fix: update react-joyride for React 18 compat

* fix: types for allowed MyInfo types

* fix: broken types assertions or code raised by linter

* fix: invalid React SVG prop clip-path -> clipPath

* fix: replace dnd package with React 18 compatible package

* fix: update package to remove console warning about deprecated calls

ReactMarkdown: Support for defaultProps will be removed from function components in a future major release.

* fix: temporary fix for menu focus color desync

final fix is to move to OGPDS

* feat: update `@testing-library/*` packages

remove unused testing-library/dom package

* feat: lock storybook packages

* fix: correctly upgrade virus-scanner type-fest package

* fix: use `isDisabled` prop over disabled

will not correctly set state if isDisabled is used instead

* fix: test assertions due to RTL changes

* fix: upgrade floating-ui package for compatibility

* fix: compat issues in ChakraUI v2

* feat: update input theming for ChakraUI v2

* fix: NumberInput render

* feat: update Drawer and Modal themes to have background using cssVars

* fix: use explicit style context required by Chakra V2

* fix: correctly import from component Button

* fix: update margin due to flex using `gap` now

instead of the margin-top previously, which allowed for margin collapse

* fix: update Rating component spacing due to flex gap change in chakra

* fix: update YesNo field margins due to Flex gap change

* fix: weird margins with collaborator menu, match width

* fix: use __css instead of sx for FeatureBanner

or text will be space-betweened

* fix: button spacing for CreateWorkspaceModal

* fix: button spacing for DeleteWorkspaceModal

* fix: button spacing for RenameWorkspaceModal

* fix: spacing on LandingPage sections

* feat: update ParagraphField story to catch multiline changes too

* feat: remove unnecessary focus-visible package

* feat: update `isTruncated` -> `noOfLines`

* feat: fix width of PermissionDropdown for consistent widths

* fix: align ViewOnlyPermission row margins

* fix: SingleSelect theme breaking because of Menu

really need to go to OGPDS soon lol

* fix: add padding to prefill lock icon

* fix: add correct msw handlers for workspace page

* fix: add back missing margins

* fix: make AvatarMenu have automatic height

instead of following button's height

* fix: add back missing menuitem padding for SingleSelect

* fix: update imported lottie file type

* fix: correct modal story color scheme for cancel button

* fix: set minH of TagInput to input's var

* fix: use input exported css variables for styling TagInput

* fix: move combobox input to the top when focused

so the border won't look janky

* feat: use mockdate decorator so skeleton width fixed in snapshot

* fix: spacing of collaborator list view again

* fix: landing page margins

* fix: avatarmenudivider margins

* fix: remove removed patches copy in dockerfile

* remove ineffective white bg on style header

* feat: upgrade typescript to 5.4.5 across app

* fix: all frontend type errors after typescript upgrade

* fix: backend type errors after upgrading typescript

---------

Co-authored-by: Ken <ken@open.gov.sg>

KenLSM committed:
Merge remote-tracking branch 'origin/develop' into 04-23-feat_add_initial_vite_config_and_new_eslint_rules

KenLSM committed:
fix: typing errors on setHasSingleSubmissionValidationError for preview

KenLSM committed:
fix: add partialdeep typing on non-en i18n

KenLSM committed:
fix: additional typing errors on stories

KenLSM committed:
test: remove global playwright timeout, use individual test timeouts

KenLSM committed:
fix: inifite mounting for StripePaymentContainer

KenLSM committed:
refactor: remove unrendered text

KenLSM committed:
fix: frm-1818 scroll lock trapping

KenLSM committed:
fix: frm-1821 folder icon focus border clashing with folder text

KenLSM committed:
fix: frm-1820 sharpen focus border n calendar

KenLSM committed:
fix: frm-1817 insufficient padding between charts and icon

KenLSM committed:
Merge remote-tracking branch 'origin/develop' into 04-23-feat_add_initial_vite_config_and_new_eslint_rules

# Conflicts:
#	frontend/.storybook/main.js
#	frontend/package-lock.json
#	frontend/package.json
#	frontend/src/components/Field/Attachment/Attachment.tsx
#	frontend/src/components/Searchbar/Searchbar.stories.tsx
#	frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/FieldListDrawer/field-panels/MyInfoPanel.tsx
#	frontend/src/features/admin-form/share/ShareFormModal.tsx
#	frontend/src/features/public-form/PublicFormContext.tsx
#	frontend/src/hooks/useMdComponents.tsx
#	playwright.config.ts
#	shared/constants/form.ts

KenLSM committed:
fix: ts typing, ts config, test config

KenLSM committed:
fix: missing variable post-merge conflict resolution

KenLSM committed:
fix: fe lint-sort

KenLSM committed:
fix: circular deps warning by rollup

KenLSM committed:
Merge remote-tracking branch 'origin/develop' into 04-23-feat_add_initial_vite_config_and_new_eslint_rules

# Conflicts:
#	frontend/.storybook/main.js

KenLSM committed:
fix: FRM-1819 add custom detection element addon flag

KenLSM committed:
try no logs

KenLSM committed:
mute all logs

KenLSM committed:
add missing tweetnacl module

KenLSM committed:
revert test code

KenLSM commented on shared/package.json:
```diff
@@ -8,6 +8,8 @@
     "libphonenumber-js": "^1.10.48",
     "lodash": "^4.17.21",
+    "tweetnacl": "^1.0.3",
+    "tweetnacl-util": "^0.15.1",
```
@kevin9foong btw we forgot to add these two packages in to `/shared` 💀. 

Finally solved final piece of why chromatic publishing [fails](https://github.com/opengovsg/FormSG/actions/runs/10843484420/job/30090747001) when using vite/rollup. `webpack` is a little looser in resolving imports thus we didn't notice it while `rollup/pnpm` are much stricter.

KenLSM committed:
Merge remote-tracking branch 'origin/develop' into 04-23-feat_add_initial_vite_config_and_new_eslint_rules

# Conflicts:
#	frontend/datadog-chunk.ts
#	frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/FieldListDrawer/FieldListOption.tsx

KenLSM committed:
fix: old imports to beautiful-dnd, outdated package-lock

KenLSM committed:
fix: linting issues from updated ts lib

KenLSM committed:
fix: additional typing issues

KenLSM committed:
fix: envvars not readable on vite builds

karrui commented:
this one got some regressions on chromatic, do you want to check?
